desc/objectives/_geometry.py

```diff
@@ -1393,3 +1393,593 @@ def compute(self, params, constants=None):
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    grid : Grid, optional
+        Collocation grid containing the nodes to evaluate at. Defaults to
+        ``LinearGrid(M=eq.M_grid, N=eq.N_grid)`` for ``Equilibrium``
+        or ``LinearGrid(M=2*eq.M, N=2*eq.N)`` for ``FourierRZToroidalSurface``.
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    _coordinates = "rtz"
+    _units = "(m^-1)"
+    _print_value_fmt = "Umbilic curvature: {:10.3e} "
+
+    def __init__(
+        self,
+        eq,
+        curve,
+        target=None,
+        bounds=None,
+        weight=1,
+        curve_grid=None,
+        equil_grid=None,
+        equil_grid_axis=None,
+        curve_fixed=False,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name="Umbilic curvature",
+    ):
+
+        if target is None and bounds is None:
+            target = -10
+
+        self._eq = curve
+        self._curve = curve
+        self._curve_grid = curve_grid
+        self._equil_grid = equil_grid
+        self._equil_grid_axis = equil_grid_axis
+        self._curve_fixed = curve_fixed
+        super().__init__(
+            things=[eq, self._curve] if not curve_fixed else [eq],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self.things[0]
+        curve = self._curve if self._curve_fixed else self.things[1]
+        # if things[1] is different than self._curve, update self._curve
+        if curve != self._curve:
+            self._curve = curve
+        if self._curve_grid is None:
+            phi_arr = jnp.linspace(0, 2 * jnp.pi, 4 * curve.N)
+            phi_arr = jnp.concatenate(
+                (phi_arr, phi_arr + 2 * jnp.pi, phi_arr + 4 * jnp.pi)
+            )
+            curve_grid = LinearGrid(
+                zeta=phi_arr, NFP_umbilic_factor=curve.NFP_umbilic_factor
+            )
+        else:
+            curve_grid = self._curve_grid
+
+        if self._equil_grid_axis is None:
+            phi_arr = jnp.linspace(0, 2 * jnp.pi, 4 * curve.N)
+            phi_arr = jnp.concatenate(
+                (phi_arr, phi_arr + 2 * jnp.pi, phi_arr + 4 * jnp.pi)
+            )
+            phi_arr = jnp.mod(phi_arr, 2 * jnp.pi)
+            equil_grid_axis = LinearGrid(
```

Comment:
what is going on here exactly? you add 2pi and 4pi and then mod it? does this only work for an umbilic factor of 3?

Reply:
this is the same thing as just repeating the original phi_arr 3 times because of the modding btw